### DISCUSSION TITLE
Using XSLT3/XPATH3

### DISCUSSION THREAD

**Discussion by stefan-jung** [dita-ot/dita-ot#4084] [Ideas]
Hi, is it "safe" to upgrade plugins to XSLT3/XPATH3? It's out since 2017. The coding conventions don't mention it. Are there any Saxon licensing issues or limitations? Is it ok to mix XSLT2 files with XSLT3 files? https://www.dita-ot.org/dev/topics/plugin-coding-conventions.html

**Comment by chrispy-snps** [comment]
@xephon2 - XSLT3 has been working great for me.

I have been using XSLT3 in my plugin stylesheets for over a year now. They integrate seamlessly into the DITA-OT's XSLT2 processing, both with the open-source DITA-OT (Saxon HE) and with Syncro Soft's DITA-OT (Saxon EE).

The primary XSLT3 features I have been using for DITA-OT plugins are:

* Having a cleaner way to define a template mode that defaults to an identity transformation:

    ```
    <!-- default to identity transformation -->
    <xsl:mode name="snps-add-endnote-number" on-no-match="shallow-copy"/>
    
    <!-- modify only this element -->
    <xsl:template match="*[contains-token(@class, 'snps-d/endnote')]" mode="snps-add-endnote-number">
      ...
    </xsl:template>
    ```

* Using `contains-token()` to check for `@class` values:

    ```
    <xsl:template match="*[contains-token(@class, 'topic/note')]">
      ...
    </xsl:template>
    ```

* String concatenation using `||` instead of `concat()`:

    ```
    <xsl:with-param name="id" select="'Figure-' || local-name()"/>
    ```
There is some more good stuff mentioned here that I hope to start using:

[What’s new in XSLT 3.0 and XPath 3.1?](http://dh.obdurodon.org/xslt3.xhtml)

**Reply by stefan-jung** [reply]
Great, thank you, @chrispy-snps. I am also escpecially interested in the new string concatenation syntax.
